### PR TITLE
Rework path handling

### PR DESCRIPTION
**Pull request by lephilousophe**
So here it is.

This PR does a major overhaul of our Path handling and tries to use Path everywhere it's possible.
This should avoid path separator mixing on Windows make things cleaner to use.
For example, all of the Archives now use Path underneath instead of converting Path to String at every lookup.

There are a bunch of commits:
- first some Common code additions on String as it's the underlying type of Path
- then, Path class is reworked to improve its performance and avoid copies and inline what it can
- in addition, a new Path widget is introduced to display Path instead of using a classic label: this handles placeholder case when path is empty,
- finally all parts of ScummVM are migrated to the Path object starting with base code and continuing up to engines and GUI.
- last commits must not be merged, they are only there to catch cases where Path conversion and toString calls are done and helped to spot parts of code still needing conversion.

About engines, I tried to convert the most I could but it's not always easy because all paths ultimately begin their life as a string.
I considered that all `char *` strings don't need conversion because are hardcoded strings (this is not always true as we sometimes use `String::c_str()` to pass paths around but this can be tackled later).
The Path objects are used when an engine code uses File based stuff and I tried to spread the Path up to a point where it's not really a Path anymore but that is not perfect: some engines have a clear demarcation when interacting with the filesystem (like path conversion function) and some are fuzzier.

This will obviously need a lot of testing and review from people familiar with every part of the code.
The build has not yet been tested on all of our supported platforms neither.

I feel we miss some Path functions to mangle file extensions, it's a pattern I saw a lot in our code and I used suboptimal ways to do it. I still plan to have it done before the merge.

### CODE REVIEW

**Comment by lephilousophe** [comment]
Sorry for all the noise, the PR now builds fine on my Buildbot.

**Comment by lephilousophe** [comment]
Rebasing over last week changes.

**Comment by lephilousophe** [comment]
Rebasing over last week changes.

**Comment by lephilousophe** [comment]
Updated the PR taking latest changes into account.
I still need to cleanup the last commits which are not intended to merge

**Comment by sev-** [comment]
Wheeeeee!!!!!

**Comment by mikrosk** [comment]
This PR has introduced a performance regression, see https://bugs.scummvm.org/ticket/14853 for details.

**Comment by lephilousophe** [comment]
I issued #5621 to try to tackle this.
Thanks for the investigation.

File: common/file.cpp [hunk 1]
```diff
@@ -150,24 +150,39 @@ DumpFile::~DumpFile() {
...
+		// If the parent directory already exists keep it simple
+		AbstractFSNode *node = g_system->getFilesystemFactory()->makeFileNodePath(dirname.toString(Common::Path::kNativeSeparator));
+		if (!node->exists()) {
+			delete node;
+
+			dirname = dirname.normalize();
+			StringArray components = dirname.splitComponents();
+
+			Common::Path subpath;
+			for (StringArray::iterator it = components.end() - 1; it != components.begin(); --it) {
```
[the remark this answers is not in the document]
Comment (lephilousophe) [reply common/file.cpp:169]: Thanks, I have fixed it.
It has been tested on Linux.

**Pull request merged**